datastore/longhorn.go

```diff
@@ -1227,6 +1261,43 @@ func (s *DataStore) ListNodesRO() ([]*longhorn.Node, error) {
 	return s.nLister.Nodes(s.namespace).List(labels.Everything())
 }
 
+func (s *DataStore) ListNodesHaveEngineImage(image string) (map[string]*longhorn.Node, error) {
+	nodes, err := s.ListNodes()
+	if err != nil {
+		return nil, err
+	}
+	eiName := types.GetEngineImageChecksumName(image)
+	ei, err := s.GetEngineImage(eiName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get engine image %v: %v", image, err)
+	}
+	if ei.Status.State != types.EngineImageStateDeployed && ei.Status.State != types.EngineImageStateDeploying {
+		return nil, fmt.Errorf("engine image %v (%v) is in state %v", ei.Name, image, ei.Status.State)
+	}
+
+	for nodeID := range nodes {
+		if !ei.Status.NodeDeploymentMap[nodeID] {
+			delete(nodes, nodeID)
+		}
+	}
+	return nodes, nil
+}
+
+// ListReadyNodesHaveEngineImage returns list of ready nodes that have the corresponding engine image deployed
+func (s *DataStore) ListReadyNodesHaveEngineImage(image string) (map[string]*longhorn.Node, error) {
+	nodes, err := s.ListNodesHaveEngineImage(image)
+	if err != nil {
+		return nil, err
+	}
+	for nodeID, node := range nodes {
+		nodeReadyCondition := types.GetCondition(node.Status.Conditions, types.NodeConditionTypeReady)
+		if nodeReadyCondition.Status != types.ConditionStatusTrue {
+			delete(nodes, nodeID)
+		}
+	}
```

Comment:
No

The current convention is that `ei.Status.NodeDeploymentMap[nodeID]` is `true` when there is an EI DaemonSet pod running on the node and the container inside the pod is ready (It means that there is engine binary on the node because the container readiness probe checks for this).

Reply:
We don't use this function to update `ei.Status.NodeDeploymentMap`

Reply:
I know this function is not responsible for updating `ei.Status.NodeDeploymentMap`. I mean, if state in `ei.Status.NodeDeploymentMap[nodeID]` is not directly related to the node readiness, the `nodeReadyCondition` check is necessary. Otherwise NO.

Reply:
I see. Yes, the nodeReadyCondition check is necessary